### PR TITLE
AVRO-1985: Unreleased resources

### PR DESCRIPTION
**Pull request by gszadovszky**
Fixes for some unreleased resource issues found by static code analyzer tool.

### CODE REVIEW

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java [hunk 1]
```diff
@@ -195,25 +195,31 @@ public boolean isFlushOnEveryBlock() {
...
+      if (codecBytes != null) {
+        String strCodec = new String(codecBytes, "UTF-8");
+        this.codec = CodecFactory.fromString(strCodec).createInstance();
+      } else {
+        this.codec = CodecFactory.nullCodec().createInstance();
+      }
+    } finally {
+      if (reader != null)
+        reader.close();
+      in.close();
```
Comment (busbey) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: Why the change to close the passed SeekableInput? Shouldn't that be up to the caller?

File: lang/java/tools/src/main/java/org/apache/avro/tool/IdlTool.java [hunk 2]
```diff
@@ -58,12 +58,17 @@ public int run(InputStream in, PrintStream out, PrintStream err,
       parser = new Idl(in);
     }
 
-    if (args.size() == 2 && ! "-".equals(args.get(1))) {
-      parseOut = new PrintStream(new FileOutputStream(args.get(1)));
-    }
+      if (args.size() == 2 && !"-".equals(args.get(1))) {
```
Comment (busbey) [inline lang/java/tools/src/main/java/org/apache/avro/tool/IdlTool.java]: it looks like these indents are incorrect.

File: lang/java/tools/src/main/java/org/apache/avro/tool/Main.java [hunk 3]
```diff
@@ -104,18 +104,26 @@ private int run(String[] args) throws Exception {
 
   private static void printStream(InputStream in) throws Exception {
     byte[] buffer = new byte[1024];
-    for (int i = in.read(buffer); i != -1; i = in.read(buffer))
-      System.err.write(buffer, 0, i);
+    try {
+      for (int i = in.read(buffer); i != -1; i = in.read(buffer))
+        System.err.write(buffer, 0, i);
+    } finally {
+      in.close();
```
Comment (busbey) [inline lang/java/tools/src/main/java/org/apache/avro/tool/Main.java]: Why are we closing in within this method instead of having the caller do it?

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 1 above)
Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: You're right, it should be closed by the caller but it is already implemented this way as we are closing the reader which closes the SeekableInput as well. I did not want to break backward compatibility but fix the actual situation that the resources are not closed every time.
That's why I've tried to highlight in the method comment that this method closes the specified SeekableInput.

File: lang/java/tools/src/main/java/org/apache/avro/tool/IdlTool.java (hunk 2 above)
Comment (gszadovszky) [inline lang/java/tools/src/main/java/org/apache/avro/tool/IdlTool.java]: Seems you're right. I'll check it out.

File: lang/java/tools/src/main/java/org/apache/avro/tool/Main.java (hunk 3 above)
Comment (gszadovszky) [inline lang/java/tools/src/main/java/org/apache/avro/tool/Main.java]: This one is a private method and it seemed to be easier and shorter to close the stream inside this method instead of the public caller. Do you think I should rewrite it?

**Comment by gszadovszky** [comment]
@busbey, could you check my answers, please?

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 1 above)
Comment (busbey) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: Can we update DataFileReader to properly not close the passed input stream instead?

File: lang/java/tools/src/main/java/org/apache/avro/tool/Main.java (hunk 3 above)
Comment (busbey) [inline lang/java/tools/src/main/java/org/apache/avro/tool/Main.java]: yeah please have the caller close it instead. we should be internally disciplined.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 1 above)
Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: We can but it would be a backward incompatible change as this method is public. 
If we would like to follow the usual pattern for this case we would need another method which does not close it and deprecate this one then remove this one in the next major release.
Or is it enough marking this JIRA with "incompatible change" and change as you are suggesting?

File: lang/java/tools/src/main/java/org/apache/avro/tool/Main.java (hunk 3 above)
Comment (gszadovszky) [inline lang/java/tools/src/main/java/org/apache/avro/tool/Main.java]: Fair enough. On the way...

**Comment by gszadovszky** [comment]
@busbey, could you please check the latest changes?

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java [hunk 4]
```diff
@@ -190,11 +190,17 @@ public boolean isFlushOnEveryBlock() {
 
   /** Open a writer appending to an existing file. */
   public DataFileWriter<D> appendTo(File file) throws IOException {
-    return appendTo(new SeekableFileInput(file),
-                    new SyncableFileOutputStream(file, true));
+    SeekableInput input = new SeekableFileInput(file);
+    OutputStream output = new SyncableFileOutputStream(file, true);
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: I think this line should be inside the try block as well to guarantee that input is always closed. Or, alternatively, it may be before the first line if order does not matter.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java [hunk 5]
```diff
@@ -190,11 +190,17 @@ public boolean isFlushOnEveryBlock() {
...
+    try {
+      return appendTo(input, output);
+    } finally {
+      input.close();
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: Could you please add a comment about why output does not have to be closed?

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java [hunk 6]
```diff
@@ -190,11 +190,17 @@ public boolean isFlushOnEveryBlock() {
...
+    }
   }
 
   /** Open a writer appending to an existing file.
+   * <strong>Since 1.9.0 this method does not close in.</strong>
```
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: 1.9.0 is already out. Additionally, isn't this an API breaking change?

File: lang/java/avro/src/main/java/org/apache/avro/data/Json.java [hunk 7]
```diff
@@ -55,8 +56,12 @@ private Json() {}                               // singleton: no public ctor
   public static final Schema SCHEMA;
   static {
     try {
```
Comment (nandorKollar) [inline lang/java/avro/src/main/java/org/apache/avro/data/Json.java]: Do we have to support Java 1.6 in Avro, or we require 1.7+? I'm wondering if we can use try-with-resource statements instead of closing the stream in finally. Also, can we merge the two nested try{} blocks into one?

File: lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java [hunk 8]
```diff
@@ -270,17 +270,22 @@ File writeToDestination(File src, File destDir) throws IOException {
...
+          fw = new FileWriter(f);
+        }
         fw.write(FILE_HEADER);
         fw.write(contents);
       } finally {
-        fw.close();
+        if (fw != null)
+          fw.close();
+        if (fos != null)
+          fos.close();
```
Comment (zivanfi) [inline lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java]: It seems to me that the second close() may not be called if the first one throws an IOException.

File: lang/java/avro/src/main/java/org/apache/avro/data/Json.java (hunk 7 above)
Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/data/Json.java]: Currently, Avro supports 1.6 so we are not able to use try-with-resources, unfortunately.

File: lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java (hunk 8 above)
Comment (gszadovszky) [inline lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java]: fos.close() is invoked separately to close the file in case of the fw creation fails (fw starts writing at the constructor). If both fos and fw creation was successful than a problem occurs during writing fw.close() shall close the file. If it fails, fos.close() would probably fail too.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 6 above)
Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: Yes, unfortunately, this changes didn't make 1.9.0. And yes, this is an API breaking change so the next target is 1.10.0. Will rewrite the comment.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 5 above)
Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: good point. Will do it.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 6 above)
Comment (busbey) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: 1.9.0 isn't out yet afaik, so this should be fine as is.

File: lang/java/avro/src/main/java/org/apache/avro/data/Json.java (hunk 7 above)
Comment (busbey) [inline lang/java/avro/src/main/java/org/apache/avro/data/Json.java]: We could start discussing dropping 1.6 support in the 1.9.0 release on the dev list.

File: lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java (hunk 6 above)
Comment (zivanfi) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: Ah, sorry, I was thinking of Parquet... :(

Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/file/DataFileWriter.java]: Yes, that was my mistake too :)

**Comment by gszadovszky** [comment]
Pushed another update (hopefully, the last one).
@busbey, could you please check if this one suits you, so I can commit it.
Thanks a lot in advance.
